packages/google_sign_in/android/src/main/java/io/flutter/plugins/googlesignin/GoogleSignInPlugin.java

```diff
@@ -488,147 +392,104 @@ private void finishWithError(String errorCode, String errorMessage) {
+                result.success(tokenResult);
+              } catch (ExecutionException e) {
+                if (e.getCause() instanceof UserRecoverableAuthException) {
+                  if (shouldRecoverAuth) {
+                    // Move this to top of getTokens method.
+                    checkAndSetPendingOperation(METHOD_GET_TOKENS, result, email);
+                    recoverAuthFromException((UserRecoverableAuthException) e.getCause());
+                  } else {
+                    result.error(ERROR_USER_RECOVERABLE_AUTH, e.getLocalizedMessage(), null);
+                  }
+                } else {
+                  result.error(ERROR_REASON_EXCEPTION, e.getCause().getMessage(), null);
+                }
+              } catch (InterruptedException e) {
+                result.error(ERROR_REASON_EXCEPTION, e.getMessage(), null);
+                Thread.currentThread().interrupt();
+              }
+            }
+          });
+    }
 
-      /**
-       * Invoked when the GMS client has successfully connected to the GMS server. This signals that
-       * this listener is properly initialized.
-       */
-      @Override
-      public void onConnected(Bundle connectionHint) {
-        // We can get reconnected if, e.g. the activity is paused and resumed.
-        if (pendingOperation != null && pendingOperation.method.equals(METHOD_INIT)) {
-          finishWithSuccess(null);
-        }
+    @Override
+    public boolean onActivityResult(int requestCode, int resultCode, Intent data) {
+      if (pendingOperation == null) {
+        return false;
       }
-
-      /**
-       * Invoked when the GMS client was unable to connect to the GMS server, either because of an
-       * error the user was unable to resolve, or because the user canceled the resolution (e.g.
-       * cancelling a dialog instructing them to upgrade Google Play Services). This signals that we
-       * were unable to properly initialize this listener.
-       */
-      @Override
-      public void onConnectionFailed(@NonNull final ConnectionResult result) {
-        if (resolvingError) {
-          // Already attempting to resolve an error.
-          return;
-        } else if (result.hasResolution() && registrar.activity() != null) {
-          resolvingError = true;
-          try {
-            result.startResolutionForResult(registrar.activity(), REQUEST_CODE_RESOLVE_ERROR);
-          } catch (SendIntentException e) {
-            resolvingError = false;
-            finishWithError(ERROR_REASON_CONNECTION_FAILED, String.valueOf(result.getErrorCode()));
+      switch (requestCode) {
+        case REQUEST_CODE_RECOVER_AUTH:
+          if (resultCode == Activity.RESULT_OK) {
+            // Recover the previous result and data and attempt to get tokens again.
+            Result result = pendingOperation.result;
+            String email = (String) pendingOperation.data;
+            pendingOperation = null;
+            getTokens(result, email, false);
```

Comment:
It is also worth noting that this operation might result in "Can't recover auth" error even with recoverAuth set to true. There's not much we can do about that unless we can distinguish this second call and change the error message.